device/platforms/stm32/GPIODevice.h

```diff
@@ -0,0 +1,100 @@
+/**
+ * HAL Implementation for GPIO Devices
+ *
+ * @author Aaron Chan
+ */
+
+#ifndef LAUNCH_CORE_GPIODEVICE_H
+#define LAUNCH_CORE_GPIODEVICE_H
+
+#include "stm32f4xx_hal_gpio.h"
+#include "stm32f4xx_hal_def.h"
+#include "stm32f446xx.h"
+
+
+#include "device/GPIODevice.h"
+#include "HAL_Handlers.h"
+#include "sched/sched.h"
+#include "sync/BlockingSemaphore.h"
+
+class HALGPIODevice : public GPIODevice, public CallbackDevice {
+public:
+
+    HALGPIODevice(const char *name, GPIO_TypeDef *halGPIO, uint16_t pin) : halGPIO(halGPIO),
+                                                                           pin(pin),
+                                                                           currentBlocked(-1),
+                                                                           taskLock(1),
+                                                                           GPIODevice(name) {};
+
+
+    RetType init() override {
+        return HALHandlers::register_gpio(halGPIO, this, REG_NUM);
+    }
+
+    RetType obtain() override {
+        return RET_SUCCESS;
+    }
+
+    RetType poll() override {
+        return RET_SUCCESS;
+    }
+
+    RetType release() override {
+        return RET_SUCCESS;
+    }
+
+    RetType set(uint32_t val) override {
+        RESUME();
+
+        RetType ret = CALL(taskLock.acquire());
+        RET_CHECK(ret);
+
+        taskLock = sched_dispatched;
+
+        HAL_GPIO_WritePin(halGPIO, this->pin, val);
```

Comment:
may want some kind of check on val, you are casting to a HAL GPIO_PinState (which is int 0 and 1)

not sure what this will do if val != 0 or 1, probably undefined or gives you a HAL error

Reply:
also should check return type of HAL functions

Reply:
WritePin is a void function